In [2]:
import pandas as pd

In [4]:
df = pd.read_csv('friends.csv')
df

fname    lname  age sex section height(cm)  weight(kg)  spend_A  \
0     Robel   Petros   37   M       B      175.4        85.6   1500.0   
1     Rakeb   Kahlid   38   M       B        NaN         NaN      NaN   
2    Nitsuh   Moreda   35   M       B      165.1         NaN    500.0   
3      Aron    Dawit   40   M       A          0        55.6    300.0   
4    Nardos     Sema   39   M       B      160.8       160.0    200.0   
5   Solomon   Meftir   36   M       A         xx       -60.0    100.0   
6  Salahdin    dawud   35   M       B      170.3        75.5      NaN   
7    Soresa    Dawud   35   M       C      155.1        56.8   1000.0   
8   Sewunet  Kahilid   36   M       C      154.2        60.6    600.0   
9      Aman   Vamrud   38   M       A      162.1        58.7    600.0   

   spend_B  spend_C  
0    200.0    300.0  
1   1000.0    200.0  
2   -100.0    200.0  
3    300.0      NaN  
4      NaN    300.0  
5    400.0      NaN  
6    500.0    500.0  
7      NaN      NaN  
8    300.0    200.0  
9    200.0    200.0

In [8]:
df.dtypes

fname          object
lname          object
age             int64
sex            object
section        object
height(cm)     object
weight(kg)    float64
spend_A       float64
spend_B       float64
spend_C       float64
dtype: object

**Handling invalid values**

> There is an invalid value coded as "xx" in the height(cm) column. We can coerce invalid values to "NaN" using the errors keyword as follows:-

In [14]:
df["height(cm)"] = pd.to_numeric(df["height(cm)"], errors='coerce')

# df["height(cm)"] = df["height(cm)"].astype(float, errors = 'ignore')
df

fname    lname  age sex section  height(cm)  weight(kg)  spend_A  \
0     Robel   Petros   37   M       B       175.4        85.6   1500.0   
1     Rakeb   Kahlid   38   M       B         NaN         NaN      NaN   
2    Nitsuh   Moreda   35   M       B       165.1         NaN    500.0   
3      Aron    Dawit   40   M       A         0.0        55.6    300.0   
4    Nardos     Sema   39   M       B       160.8       160.0    200.0   
5   Solomon   Meftir   36   M       A         NaN       -60.0    100.0   
6  Salahdin    dawud   35   M       B       170.3        75.5      NaN   
7    Soresa    Dawud   35   M       C       155.1        56.8   1000.0   
8   Sewunet  Kahilid   36   M       C       154.2        60.6    600.0   
9      Aman   Vamrud   38   M       A       162.1        58.7    600.0   

   spend_B  spend_C  
0    200.0    300.0  
1   1000.0    200.0  
2   -100.0    200.0  
3    300.0      NaN  
4      NaN    300.0  
5    400.0      NaN  
6    500.0    500.0  
7      NaN      NaN  
8    300.0    200.0  
9    200.0    200.0

In [16]:
df["height(cm)"].dtype

dtype('float64')

In [18]:
df.dtypes

fname          object
lname          object
age             int64
sex            object
section        object
height(cm)    float64
weight(kg)    float64
spend_A       float64
spend_B       float64
spend_C       float64
dtype: object

**Interpretation**
> We can see that there are discrepancies in height(cm) and weight(kg) columns.<br>
> The minimum value of height(cm) is 0. It is not possible because height cannot be 0.<br>
> The minimum and maximum values of weight(kg) are -60 and 160. Weight cannot be negative and weight cannot be as high as 160. So, both are invalid values.<br>
> They are outliers and need to be properly addressed.

In [84]:
df

fname    lname  age sex section  height(cm)  weight(kg)  spend_A  \
0     Robel   Petros   37   M       B       175.4        85.6   1500.0   
1     Rakeb   Kahlid   38   M       B         NaN         NaN      NaN   
2    Nitsuh   Moreda   35   M       B       165.1         NaN    500.0   
3      Aron    Dawit   40   M       A         0.0        55.6    300.0   
4    Nardos     Sema   39   M       B       160.8       160.0    200.0   
5   Solomon   Meftir   36   M       A         NaN       -60.0    100.0   
6  Salahdin    dawud   35   M       B       170.3        75.5      NaN   
7    Soresa    Dawud   35   M       C       155.1        56.8   1000.0   
8   Sewunet  Kahilid   36   M       C       154.2        60.6    600.0   
9      Aman   Vamrud   38   M       A       162.1        58.7    600.0   

   spend_B  spend_C  
0    200.0    300.0  
1   1000.0    200.0  
2   -100.0    200.0  
3    300.0      NaN  
4      NaN    300.0  
5    400.0      NaN  
6    500.0    500.0  
7      NaN      NaN  
8    300.0    200.0  
9    200.0    200.0

**Optional: Reorder the column labels**

In [26]:
df = df[['fname','lname','age','sex','section','height(cm)','weight(kg)','spend_A','spend_B','spend_C']]

**Dealing with negative numerical values**

> We have seen that, in the weight(kg) column, there is a negative value of -60. It is invalid value because weight cannot be negative. There is a high probability that weight is 60 kg and it is mistyped as -60. So, I will replace the negative value of -60 with positive value of 60.

We can do it as follows:-

In [20]:
df['weight(kg)'] = df['weight(kg)'].replace(-60, 60)

In [35]:
df

fname    lname  age sex section  height(cm)  weight(kg)  spend_A  \
0     Robel   Petros   37   M       B       175.4        85.6   1500.0   
1     Rakeb   Kahlid   38   M       B         NaN         NaN      NaN   
2    Nitsuh   Moreda   35   M       B       165.1         NaN    500.0   
3      Aron    Dawit   40   M       A         0.0        55.6    300.0   
4    Nardos     Sema   39   M       B       160.8       160.0    200.0   
5   Solomon   Meftir   36   M       A         NaN        60.0    100.0   
6  Salahdin    dawud   35   M       B       170.3        75.5      NaN   
7    Soresa    Dawud   35   M       C       155.1        56.8   1000.0   
8   Sewunet  Kahilid   36   M       C       154.2        60.6    600.0   
9      Aman   Vamrud   38   M       A       162.1        58.7    600.0   

   spend_B  spend_C  
0    200.0    300.0  
1   1000.0    200.0  
2   -100.0    200.0  
3    300.0      NaN  
4      NaN    300.0  
5    400.0      NaN  
6    500.0    500.0  
7      NaN      NaN  
8    300.0    200.0  
9    200.0    200.0

> Similarly, in the spend_B column, there is a negative value -100. The amount spent cannot be negative. So, we need to replace this negative value of -100 with positive value of 100.

In [88]:
df['spend_B'] = df['spend_B'].replace(-100, 100)

In [90]:
df

fname    lname  age sex section  height(cm)  weight(kg)  spend_A  \
0     Robel   Petros   37   M       B       175.4        85.6   1500.0   
1     Rakeb   Kahlid   38   M       B         NaN         NaN      NaN   
2    Nitsuh   Moreda   35   M       B       165.1         NaN    500.0   
3      Aron    Dawit   40   M       A         0.0        55.6    300.0   
4    Nardos     Sema   39   M       B       160.8       160.0    200.0   
5   Solomon   Meftir   36   M       A         NaN        60.0    100.0   
6  Salahdin    dawud   35   M       B       170.3        75.5      NaN   
7    Soresa    Dawud   35   M       C       155.1        56.8   1000.0   
8   Sewunet  Kahilid   36   M       C       154.2        60.6    600.0   
9      Aman   Vamrud   38   M       A       162.1        58.7    600.0   

   spend_B  spend_C  
0    200.0    300.0  
1   1000.0    200.0  
2    100.0    200.0  
3    300.0      NaN  
4      NaN    300.0  
5    400.0      NaN  
6    500.0    500.0  
7      NaN      NaN  
8    300.0    200.0  
9    200.0    200.0

In [26]:
df.isnull().sum()

fname         0
lname         0
age           0
sex           0
section       0
height(cm)    2
weight(kg)    2
spend_A       2
spend_B       2
spend_C       3
dtype: int64

In [34]:
import numpy as np
data = {
    'A': [1, 2, None, 4],
    'B': [11, 22, np.nan, 44],
    'C': [np.nan, 222, 444, None]
}
df = pd.DataFrame(data)
df

A     B      C
0  1.0  11.0    NaN
1  2.0  22.0  222.0
2  NaN   NaN  444.0
3  4.0  44.0    NaN

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   A       3 non-null      float64
 1   B       3 non-null      float64
 2   C       2 non-null      float64
dtypes: float64(3)
memory usage: 228.0 bytes


**Dealing with outliers**

> In the height(cm) column, there is a value of 0.0. It is not possible as height cannot be 0. So, we need to resolve it.

I will replace the 0.0 value with the mean of the height(cm) column. It can be done as follows:-

In [92]:
mean_height = df['height(cm)'].mean()
mean_height

142.875

In [94]:
mean_height = df['height(cm)'].mean()
df['height(cm)'] = df['height(cm)'].replace(0.0, mean_height)

In [96]:
df

fname    lname  age sex section  height(cm)  weight(kg)  spend_A  \
0     Robel   Petros   37   M       B     175.400        85.6   1500.0   
1     Rakeb   Kahlid   38   M       B         NaN         NaN      NaN   
2    Nitsuh   Moreda   35   M       B     165.100         NaN    500.0   
3      Aron    Dawit   40   M       A     142.875        55.6    300.0   
4    Nardos     Sema   39   M       B     160.800       160.0    200.0   
5   Solomon   Meftir   36   M       A         NaN        60.0    100.0   
6  Salahdin    dawud   35   M       B     170.300        75.5      NaN   
7    Soresa    Dawud   35   M       C     155.100        56.8   1000.0   
8   Sewunet  Kahilid   36   M       C     154.200        60.6    600.0   
9      Aman   Vamrud   38   M       A     162.100        58.7    600.0   

   spend_B  spend_C  
0    200.0    300.0  
1   1000.0    200.0  
2    100.0    200.0  
3    300.0      NaN  
4      NaN    300.0  
5    400.0      NaN  
6    500.0    500.0  
7      NaN      NaN  
8    300.0    200.0  
9    200.0    200.0

> We can see that the data value of 0.0 in height(cm) column is replaced by a proper height value.<br>
In the weight(kg) column, there is a very high absurd value of 160. It is not possible to have so much weight. Hence, it is invalid value. There is a high chance that the weight is 60 kg and it is mistakenly typed as 160 kg. So, I will replace the 160 data value with 60.

In [102]:
df['weight(kg)'] = df['weight(kg)'].replace(160.0, 60.0)

In [104]:
df

fname    lname  age sex section  height(cm)  weight(kg)  spend_A  \
0     Robel   Petros   37   M       B     175.400        85.6   1500.0   
1     Rakeb   Kahlid   38   M       B         NaN         NaN      NaN   
2    Nitsuh   Moreda   35   M       B     165.100         NaN    500.0   
3      Aron    Dawit   40   M       A     142.875        55.6    300.0   
4    Nardos     Sema   39   M       B     160.800        60.0    200.0   
5   Solomon   Meftir   36   M       A         NaN        60.0    100.0   
6  Salahdin    dawud   35   M       B     170.300        75.5      NaN   
7    Soresa    Dawud   35   M       C     155.100        56.8   1000.0   
8   Sewunet  Kahilid   36   M       C     154.200        60.6    600.0   
9      Aman   Vamrud   38   M       A     162.100        58.7    600.0   

   spend_B  spend_C  
0    200.0    300.0  
1   1000.0    200.0  
2    100.0    200.0  
3    300.0      NaN  
4      NaN    300.0  
5    400.0      NaN  
6    500.0    500.0  
7      NaN      NaN  
8    300.0    200.0  
9    200.0    200.0

**Dealing with missing numerical values**

**The following commands help us to deal with missing numerical values.**

df.isnull()

The above command checks whether each cell in a dataframe contains missing values or not. If the cell contains missing value, it returns True otherwise it returns False.

df.isnull.sum()

The above command returns the total number of missing values in each column in the dataset.

isna() and notna() functions to detect ‘NA’ values

Pandas provides isna() and notna() functions to detect ‘NA’ values. These are also methods on Series and DataFrame objects.

Examples of isna() and notna() commands

We can detect ‘NA’ values in the dataframe with the following command:-

df.isna()

We can detect ‘NA’ values in a particular column in the dataframe

pd.isna(df[‘col_name’])

df[‘col_name’].notna()

So, we can check the number of missing values in each column in the dataset as follows:-

In [108]:
df.isnull().sum()

fname         0
lname         0
age           0
sex           0
section       0
height(cm)    2
weight(kg)    2
spend_A       2
spend_B       2
spend_C       3
dtype: int64

**We can see that there are lots of missing values in the dataset.**